File: coreengine/envelope/enums.go
// Package envelope provides core enums - source of truth for jeeves-kernel.
//
// All Python enums are thin wrappers around these Go definitions.
// Proto definitions in engine.proto are used for gRPC transport only.
package envelope

// TerminalReason represents why processing terminated.
type TerminalReason string

const (
	TerminalReasonCompleted             TerminalReason = "completed"
	TerminalReasonMaxIterationsExceeded TerminalReason = "max_iterations_exceeded"
	TerminalReasonMaxLLMCallsExceeded   TerminalReason = "max_llm_calls_exceeded"
	TerminalReasonMaxAgentHopsExceeded  TerminalReason = "max_agent_hops_exceeded"
	TerminalReasonMaxLoopExceeded       TerminalReason = "max_loop_exceeded"
	TerminalReasonUserCancelled         TerminalReason = "user_cancelled"
	TerminalReasonToolFailedFatally     TerminalReason = "tool_failed_fatally"
	TerminalReasonLLMFailedFatally      TerminalReason = "llm_failed_fatally"
	TerminalReasonPolicyViolation       TerminalReason = "policy_violation"
)

// RiskLevel represents risk for tool execution.
type RiskLevel string

const (
	// Semantic levels
	RiskLevelReadOnly    RiskLevel = "read_only"
	RiskLevelWrite       RiskLevel = "write"
	RiskLevelDestructive RiskLevel = "destructive"
	// Severity levels
	RiskLevelLow      RiskLevel = "low"
	RiskLevelMedium   RiskLevel = "medium"
	RiskLevelHigh     RiskLevel = "high"
	RiskLevelCritical RiskLevel = "critical"
)

// RequiresConfirmation checks if a risk level requires user confirmation.
func (r RiskLevel) RequiresConfirmation() bool {
	return r == RiskLevelDestructive || r == RiskLevelHigh || r == RiskLevelCritical
}

// ToolCategory represents tool categorization.
type ToolCategory string

const (
	// Operation types
	ToolCategoryRead    ToolCategory = "read"
	ToolCategoryWrite   ToolCategory = "write"
	ToolCategoryExecute ToolCategory = "execute"
	ToolCategoryNetwork ToolCategory = "network"
	ToolCategorySystem  ToolCategory = "system"
	// Tool organization
	ToolCategoryUnified    ToolCategory = "unified"
	ToolCategoryComposite  ToolCategory = "composite"
	ToolCategoryResilient  ToolCategory = "resilient"
	ToolCategoryStandalone ToolCategory = "standalone"
	ToolCategoryInternal   ToolCategory = "internal"
)

// HealthStatus represents health check status.
type HealthStatus string

const (
	HealthStatusHealthy   HealthStatus = "healthy"
	HealthStatusDegraded  HealthStatus = "degraded"
	HealthStatusUnhealthy HealthStatus = "unhealthy"
	HealthStatusUnknown   HealthStatus = "unknown"
)

// LoopVerdict represents agent loop control verdicts.
type LoopVerdict string

const (
	LoopVerdictProceed  LoopVerdict = "proceed"
	LoopVerdictLoopBack LoopVerdict = "loop_back"
	LoopVerdictAdvance  LoopVerdict = "advance"
	LoopVerdictEscalate LoopVerdict = "escalate"
)

// RiskApproval represents risk approval status.
type RiskApproval string

const (
	RiskApprovalApproved RiskApproval = "approved"
	RiskApprovalDenied   RiskApproval = "denied"
	RiskApprovalPending  RiskApproval = "pending"
)

// ToolAccess represents tool access levels for agents.
type ToolAccess string

const (
	ToolAccessNone  ToolAccess = "none"
	ToolAccessRead  ToolAccess = "read"
	ToolAccessWrite ToolAccess = "write"
	ToolAccessAll   ToolAccess = "all"
)

// OperationStatus represents status of an operation result.
type OperationStatus string

const (
	OperationStatusSuccess           OperationStatus = "success"
	OperationStatusError             OperationStatus = "error"
	OperationStatusNotFound          OperationStatus = "not_found"
	OperationStatusTimeout           OperationStatus = "timeout"
	OperationStatusValidationError   OperationStatus = "validation_error"
	OperationStatusPartial           OperationStatus = "partial"
	OperationStatusInvalidParameters OperationStatus = "invalid_parameters"
)

File: jeeves_core/types/enums.py
"""Core enums - Python interface to Go kernel definitions.

Source of truth: coreengine/envelope/enums.go
"""

from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Dict, List, Optional


class RiskLevel(str, Enum):
    """Risk level for tool execution."""
    # Semantic levels
    READ_ONLY = "read_only"
    WRITE = "write"
    DESTRUCTIVE = "destructive"
    # Severity levels
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"

    @classmethod
    def requires_confirmation(cls, level: "RiskLevel") -> bool:
        """Check if a risk level requires user confirmation."""
        return level in (cls.DESTRUCTIVE, cls.HIGH, cls.CRITICAL)


class ToolCategory(str, Enum):
    """Tool categorization."""
    # Operation types
    READ = "read"
    WRITE = "write"
    EXECUTE = "execute"
    NETWORK = "network"
    SYSTEM = "system"
    # Tool organization
    UNIFIED = "unified"
    COMPOSITE = "composite"
    RESILIENT = "resilient"
    STANDALONE = "standalone"
    INTERNAL = "internal"


class HealthStatus(str, Enum):
    """Health check status."""
    HEALTHY = "healthy"
    DEGRADED = "degraded"
    UNHEALTHY = "unhealthy"
    UNKNOWN = "unknown"


class TerminalReason(str, Enum):
    """Reasons for envelope termination."""
    COMPLETED = "completed"
    MAX_ITERATIONS_EXCEEDED = "max_iterations_exceeded"
    MAX_LLM_CALLS_EXCEEDED = "max_llm_calls_exceeded"
    MAX_AGENT_HOPS_EXCEEDED = "max_agent_hops_exceeded"
    MAX_LOOP_EXCEEDED = "max_loop_exceeded"
    USER_CANCELLED = "user_cancelled"
    TOOL_FAILED_FATALLY = "tool_failed_fatally"
    LLM_FAILED_FATALLY = "llm_failed_fatally"
    POLICY_VIOLATION = "policy_violation"


class LoopVerdict(str, Enum):
    """Agent loop control verdicts."""
    PROCEED = "proceed"
    LOOP_BACK = "loop_back"
    ADVANCE = "advance"
    ESCALATE = "escalate"


class RiskApproval(str, Enum):
    """Risk approval status."""
    APPROVED = "approved"
    DENIED = "denied"
    PENDING = "pending"


class ToolAccess(str, Enum):
    """Tool access levels for agents."""
    NONE = "none"
    READ = "read"
    WRITE = "write"
    ALL = "all"


class OperationStatus(str, Enum):
    """Status of an operation result."""
    SUCCESS = "success"
    ERROR = "error"
    NOT_FOUND = "not_found"
    TIMEOUT = "timeout"
    VALIDATION_ERROR = "validation_error"
    PARTIAL = "partial"
    INVALID_PARAMETERS = "invalid_parameters"


@dataclass
class OperationResult:
    """Unified result type for operations."""
    status: OperationStatus
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None
    error_type: Optional[str] = None
    suggestions: List[str] = field(default_factory=list)

    @property
    def is_success(self) -> bool:
        """Check if operation succeeded."""
        return self.status == OperationStatus.SUCCESS

    @property
    def is_error(self) -> bool:
        """Check if operation failed."""
        return self.status != OperationStatus.SUCCESS

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization."""
        result = {"status": self.status.value}
        if self.data is not None:
            result["data"] = self.data
        if self.error is not None:
            result["error"] = self.error
        if self.error_type is not None:
            result["error_type"] = self.error_type
        if self.suggestions:
            result["suggestions"] = self.suggestions
        return result

    @classmethod
    def success(cls, data: Optional[Dict[str, Any]] = None) -> "OperationResult":
        """Create a success result."""
        return cls(status=OperationStatus.SUCCESS, data=data or {})

    @classmethod
    def error(
        cls,
        message: str,
        error_type: str = "unknown",
        suggestions: Optional[List[str]] = None,
    ) -> "OperationResult":
        """Create an error result."""
        return cls(
            status=OperationStatus.ERROR,
            error=message,
            error_type=error_type,
            suggestions=suggestions or [],
        )

    @classmethod
    def not_found(
        cls,
        message: str = "Resource not found",
        suggestions: Optional[List[str]] = None,
    ) -> "OperationResult":
        """Create a not found result."""
        return cls(
            status=OperationStatus.NOT_FOUND,
            error=message,
            error_type="not_found",
            suggestions=suggestions or [],
        )

    @classmethod
    def validation_error(
        cls,
        message: str,
        suggestions: Optional[List[str]] = None,
    ) -> "OperationResult":
        """Create a validation error result."""
        return cls(
            status=OperationStatus.VALIDATION_ERROR,
            error=message,
            error_type="validation_error",
            suggestions=suggestions or [],
        )
